framework/src/loops/AllLocalDofIndicesThread.C

```diff
@@ -21,24 +24,50 @@
 #include LIBMESH_INCLUDE_UNORDERED_SET
 LIBMESH_DEFINE_HASH_POINTERS
 
-AllLocalDofIndicesThread::AllLocalDofIndicesThread(System & sys,
+AllLocalDofIndicesThread::AllLocalDofIndicesThread(FEProblemBase & problem,
+                                                   System & sys,
                                                    std::vector<std::string> vars,
                                                    bool include_semilocal)
   : ParallelObject(sys.comm()),
+    _problem(problem),
     _sys(sys),
     _dof_map(sys.get_dof_map()),
     _vars(vars),
     _include_semilocal(include_semilocal)
 {
+  // get system base to discern between variables and array variables
+  SystemBase * system_base = dynamic_cast<SystemBase *>(&_problem.getAuxiliarySystem());
+  if (!system_base->hasVariable(vars[0]))
+    system_base = dynamic_cast<SystemBase *>(&_problem.getNonlinearSystem());
+
+  // get variable numbers
   _var_numbers.resize(_vars.size());
```

Comment:
Did you try push the numbers of all array variable components into this `_var_numbers`? The components are normal libmesh variables, so supposedly we do not need to do anything other than populating this vector properly.

Reply:
No that doesn't work. Per the slack discussion, if you use the variable numbers of any but the first array variable in dof_map, then it segfaults.

Reply:
Is it possible to check in a test here to show the problem?

Reply:
Yeah of course. Is there a transfer that works with array vars?

Reply:
Found one here: `moose/test/tests/transfers/multiapp_copy_transfer/array_variable_transfer`

Reply:
Example in ci 2. That's the error you should see when running `array.i`

```
Executing MultiApps on TIMESTEP_BEGIN
Solving MultiApp 'run_array_sub' with target time 1.000000e+00 and dt 1.000000e+00 with auto-advance off
run_array_sub0: 
run_array_sub0: Time Step 1, time = 1, dt = 1
map_find() error: key "v_aux" not found in file ../src/systems/system.C on line 1406
Stack frames: 21
```

Reply:
@snschune you have a misunderstanding on `MultiApps/run_array_sub/relaxed_variables`: it needs to be a nonlinear variable on the subapp. You set `v_aux` here which probably should be `v`. I tried just passing `FEProblemBase` in `AllLocalDofIndicesThread`, which seems ok. I just attached my changes there:
```
diff --git a/framework/src/loops/AllLocalDofIndicesThread.C b/framework/src/loops/AllLocalDofIndicesThread.C
index 3e5d61b279..24f3e0add9 100644
--- a/framework/src/loops/AllLocalDofIndicesThread.C
+++ b/framework/src/loops/AllLocalDofIndicesThread.C
@@ -11,6 +11,8 @@
 
 #include "FEProblem.h"
 #include "ParallelUniqueId.h"
+#include "NonlinearSystemBase.h"
+#include "MooseVariableFE.h"
 
 #include "libmesh/dof_map.h"
 #include "libmesh/threads.h"
@@ -21,24 +23,35 @@
 #include LIBMESH_INCLUDE_UNORDERED_SET
 LIBMESH_DEFINE_HASH_POINTERS
 
-AllLocalDofIndicesThread::AllLocalDofIndicesThread(System & sys,
+AllLocalDofIndicesThread::AllLocalDofIndicesThread(FEProblemBase & problem,
                                                    std::vector<std::string> vars,
                                                    bool include_semilocal)
-  : ParallelObject(sys.comm()),
-    _sys(sys),
-    _dof_map(sys.get_dof_map()),
+  : ParallelObject(problem.comm()),
+    _problem(problem),
+    _sys(problem.getNonlinearSystemBase().system()),
+    _dof_map(_sys.get_dof_map()),
     _vars(vars),
     _include_semilocal(include_semilocal)
 {
-  _var_numbers.resize(_vars.size());
   for (unsigned int i = 0; i < _vars.size(); i++)
-    _var_numbers[i] = _sys.variable_number(_vars[i]);
+  {
+    auto & var = _problem.getVariable(0, _vars[i]);
+    if (var.count() > 1) // array variable
+    {
+      const auto & array_var = _problem.getArrayVariable(0, _vars[i]);
+      for (unsigned int p = 0; p < var.count(); ++p)
+        _var_numbers.push_back(_sys.variable_number(array_var.componentName(p)));
+    }
+    else
+      _var_numbers.push_back(_sys.variable_number(_vars[i]));
+  }
 }
```